### DISCUSSION TITLE
Podman warns about cgroups-v1 being deprecated, but don't provide more info on how to migrate to cgroup v2

### DISCUSSION THREAD

**Discussion by fishy** [containers/podman#23559] [General]
### Issue Description

When I run any podman command, it will warn about cgroups-v1 being deprecated in favor of cgroups-v2, but it does not provide any more info on how do I switch to cgroups-v2 (it does provide info about how to silent that warning):
```
$ podman version
WARN[0000] Using cgroups-v1 which is deprecated in favor of cgroups-v2 with Podman v5 and will be removed in a future version. Set environment variable `PODMAN_IGNORE_CGROUPSV1_WARNING` to hide this warning. 
Client:       Podman Engine
Version:      5.0.3
API Version:  5.0.3
Go Version:   go1.22.5
Built:        Wed Dec 31 16:00:00 1969
OS/Arch:      linux/amd64
```

I checked my `~/.config/containers/containers.conf` file, and every reference of `cgroup` I can find there are in comments, which seems to suggest that everything related to cgroups are being configured to use the default config to me? I also don't have `~/.config/containers/storage.conf` so that's default as well.

And it seems to me that my system supports both v1 and v2 of cgroups:

```
$ grep cgroup /proc/filesystems
nodev   cgroup
nodev   cgroup2
```

So if my system supports cgroups v2 and I did not explicitly configure podman to use cgroups v1 and cgroups v1 is deprecated in podman, it should auto pick v2 instead of warn me about that?

### Steps to reproduce the issue

Steps to reproduce the issue
1.
2.
3.

### Describe the results you received

Describe the results you received

### Describe the results you expected

Describe the results you expected

### podman info output

```yaml
$ podman info
WARN[0000] Using cgroups-v1 which is deprecated in favor of cgroups-v2 with Podman v5 and will be removed in a future version. Set environment variable `PODMAN_IGNORE_CGROUPSV1_WARNING` to hide this warning. 
host:
  arch: amd64
  buildahVersion: 1.35.3
  cgroupControllers: []
  cgroupManager: cgroupfs
  cgroupVersion: v1
  conmon:
    package: conmon_2.1.10+ds1-1+b1_amd64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.10, commit: unknown'
  cpuUtilization:
    idlePercent: 98.51
    systemPercent: 0.28
    userPercent: 1.21
  cpus: 8
  databaseBackend: boltdb
  distribution:
    codename: trixie
    distribution: debian
    version: unknown
  eventLogger: journald
  freeLocks: 2047
  hostname: penguin
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  kernel: 6.6.30-02726-gb84bbdf6955e
  linkmode: dynamic
  logDriver: journald
  memFree: 10522427392
  memTotal: 15184396288
  networkBackend: netavark
  networkBackendInfo:
    backend: netavark
    dns:
      package: aardvark-dns_1.4.0-6_amd64
      path: /usr/lib/podman/aardvark-dns
      version: aardvark-dns 1.4.0
    package: netavark_1.4.0-4.1_amd64
    path: /usr/lib/podman/netavark
    version: netavark 1.4.0
  ociRuntime:
    name: /home/fishy/bin/runsc-podman.sh
    package: Unknown
    path: /home/fishy/bin/runsc-podman.sh
    version: |-
      runsc version release-20240729.0
      spec: 1.1.0-rc.1
  os: linux
  pasta:
    executable: /usr/bin/pasta
    package: passt_0.0~git20240726.57a21d2-1_amd64
    version: |
      pasta 0.0~git20240726.57a21d2-1
      Copyright Red Hat
      GNU General Public License, version 2 or later
        <https://www.gnu.org/licenses/old-licenses/gpl-2.0.html>
      This is free software: you are free to change and redistribute it.
      There is NO WARRANTY, to the extent permitted by law.
  remoteSocket:
    exists: true
    path: /run/user/1000/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns_1.2.1-1+b1_amd64
    version: |-
      slirp4netns version 1.2.1
      commit: 09e31e92fa3d2a1d3ca261adaeb012c8d75a8194
      libslirp: 4.8.0
      SLIRP_CONFIG_VERSION_MAX: 4
      libseccomp: 2.5.5
  swapFree: 0
  swapTotal: 0
  uptime: 155h 40m 3.00s (Approximately 6.46 days)
  variant: ""
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  - ipvlan
  volume:
  - local
registries: {}
store:
  configFile: /home/fishy/.config/containers/storage.conf
  containerStore:
    number: 1
    paused: 0
    running: 1
    stopped: 0
  graphDriverName: vfs
  graphOptions: {}
  graphRoot: /home/fishy/.local/share/containers/storage
  graphRootAllocated: 136902082560
  graphRootUsed: 112356233216
  graphStatus: {}
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 5
  runRoot: /run/user/1000/containers
  transientStore: false
  volumePath: /home/fishy/.local/share/containers/storage/volumes
version:
  APIVersion: 5.0.3
  Built: 0
  BuiltTime: Wed Dec 31 16:00:00 1969
  GitCommit: ""
  GoVersion: go1.22.5
  Os: linux
  OsArch: linux/amd64
  Version: 5.0.3
```
```


### Podman in a container

Yes

### Privileged Or Rootless

Rootless

### Upstream Latest Release

No

### Additional environment details

I'm using podman distributed by Debian testing (trixie), so it's not the latest release.

### Additional information

Additional information like issue happens only occasionally or issue happens with a particular architecture or on a particular setting

**Comment by ScrambledBrain** [comment]
I agree, that the pre-requisites / requirements are not documented as they could / should be.

For example such links should contain more information about cgroups:
https://github.com/containers/podman/blob/main/docs/tutorials/podman_tutorial.md
https://github.com/containers/podman/blob/main/docs/tutorials/rootless_tutorial.md

Or maybe another "tutorial" link about OS configuration would help, which should contain the pre-requisites / requirements for rootless / rootful operating modes of podman with all the dis-/advantages [cgroups v1 vs. v2, sysctl settings, ...] in a nice table as a good overview.

I also only stumpled over the cgroups topic by the small sentence somewhere in the long text of [Systemd Units via Quadlet](https://docs.podman.io/en/latest/markdown/podman-systemd.unit.5.html) :
> Quadlet requires the use of cgroup v2, use podman info --format {{.Host.CgroupsVersion}} to check on the system.

**Comment by afbjorklund** [comment]
Strange that you didn't get systemd (with cgroups v2) and weird that it uses VFS (not overlayfs)

```
  cgroupControllers: []
  cgroupManager: cgroupfs
  cgroupVersion: v1

  graphDriverName: vfs
  graphStatus: {}
```

The default settings for my Podman 5.2.0 on Debian 12 (built from source code) ended up as:

```
  cgroupControllers:
  - cpu
  - pids
  cgroupManager: systemd
  cgroupVersion: v2

  graphDriverName: overlay
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Supports shifting: "false"
    Supports volatile: "true"
    Using metacopy: "false"
```

**Reply by jaknz** [reply]
> Strange that you didn't get systemd (with cgroups v2) and weird that it uses VFS (not overlayfs)
> 
> ```
>   cgroupControllers: []
>   cgroupManager: cgroupfs
>   cgroupVersion: v1
> 
>   graphDriverName: vfs
>   graphStatus: {}
> ```

I'm running podman on WSL2 in Windows 11, and I'm getting this as well. Definitely do have cgroups v2 on my up-to-date Debian, but I'm wondering if we're getting pulled back to v1 because of the way the WSL2 kernel mounts the VFS?

**Reply by fishy** [reply]
I'm running podman on Crostini which is similar to WSL so I guess that's also why I'm seeing this (no cgroup v2 available).

**Comment by afbjorklund** [comment]
Debian changed to cgroups v2 by default, in Debian 11 "bullseye":

https://www.debian.org/releases/bullseye/amd64/release-notes/ch-whats-new.en.html#cgroupv2

For the older versions, it had to be enabled on the kernel cmdline:

`systemd.unified_cgroup_hierarchy=true`

**Comment by jaknz** [comment]
Found the solution, at least for WSL2!
https://github.com/spurin/wsl-cgroupsv2

**Reply by symplrdudley** [reply]
This looks promising!

**Comment by pavelanni** [comment]
I got this warning after upgrading from RHEL 9.4 to 9.5 in which Podman was upgraded from 4.9 to 5.2. I checked with `podman info` and indeed, it used groups v1. Probably because I kept updating this server for several years already and inherited this setting from pervious versions.

After searching for a while, I found this command and it solved the issue!

`sudo grubby --update-kernel=ALL --args="systemd.unified_cgroup_hierarchy=1"`

For reference, it was on Oracle's doc site: https://docs.oracle.com/en/learn/ol-cgroup-v2/#enable-control-group-version-2

**Comment by mirabilos** [comment]
And what do I set on Debian trixie to not use cgroups (due to using sysvinit) with crun?

cf. https://github.com/containers/crun/issues/433#issuecomment-4982075655

**Comment by mirabilos** [comment]
Huh… I mounted `-t cgroup2 none /sys/fs/cgroup` and it *still* shows me that error:

```
$ podman run --rm hello-world
WARN[0000] Using cgroups-v1 which is deprecated in favor of cgroups-v2 with Podman v5 and will be removed in a future version. Set environment variable `PODMAN_IGNORE_CGROUPSV1_WARNING` to hide this warning.
Error: OCI runtime error: crun: invalid file system type on `/sys/fs/cgroup`
```

**Reply by mirabilos** [reply]
In case anyone else runs into this, I managed to fix this purely by accident. Three one-line changes were needed:

① if `mount | fgrep cgroup` shows nothing: add to `/etc/fstab` then `sudo mount -a`

    none /sys/fs/cgroup cgroup2 defaults 0 0

② if such a directory is not automatically created, add to `/etc/rc.local` (where `pott` is the user under which one wants to run pottman and crun):

    install -d -o pott -g pott -m 0700 /tmp/runtime-pott

③ if such variable is not automatically set at login, add to `~pott/.profile`:

    XDG_RUNTIME_DIR=/tmp/runtime-pott; export XDG_RUNTIME_DIR

Symptom for ① was:

```
WARN[0000] Using cgroups-v1 which is deprecated in favor of cgroups-v2 with Podman v5 and will be removed in a future version. Set environment variable `PODMAN_IGNORE_CGROUPSV1_WARNING` to hide this warning. 
Error: OCI runtime error: crun: invalid file system type on `/sys/fs/cgroup`
```

Symptom for ②/③ was:

```
pott@tgb1:~ $ (unset XDG_RUNTIME_DIR; crun list)
2026-07-19T18:26:26.868714Z: create directory `/run/crun`: No such file or directory
1|pott@tgb1:~ $ crun list
NAME PID       STATUS   BUNDLE PATH                             CREATED                        OWNER
```

Basically, without that magic environment variable set, it did not do a suitable fallback but instead guessed it was run as `root`, for some reason.

----

And because apparently, “what systemd does” can be expected 🤬, this command was also needed in `/etc/rc.local`:

    mount --make-rshared /

Symptom for *that* was:

    WARN[0000] "/" is not a shared mount, this could cause issues or missing mounts with rootless containers
